### ISSUE TITLE
Add support for Steam Family Games

### ISSUE THREAD

**Issue by balu100** [JosefNemec/PlayniteExtensions#396]
#### Description:
Add functionality to retrieve and list Steam family games using the Steam API. This will involve integrating with the `GetFamilyGroup` and `GetSharedLibraryApps` endpoints provided by the Steam API.

#### API Documentation:
- [IFamilyGroupsService/GetFamilyGroup](https://steamapi.xpaw.me/#IFamilyGroupsService/GetFamilyGroup) - needed for retrieving the Group ID.
- [IFamilyGroupsService/GetSharedLibraryApps](https://steamapi.xpaw.me/#IFamilyGroupsService/GetSharedLibraryApps) - group ID is required even though the documentation says it's not.

**Comment by Shaarigan** [comment]
+1 From my End. Playnite makes it easier to have an overview about all the games by looking into 2 instead of 8 clients to oversee the library but this lacks of features like selecting a random game from games that are shared with my account

**Comment by EpiKnightz** [comment]
I tried using the API above but still seem like it needs private API key of each family member. Even though it can get family_groupid, can get steamid of each member, at the final steps it just returns blank if that member game list is private.

For now I guess the best option is using Additional Accounts settings in Steam Integration, or ask your members profile to change to public and call GetLibraryGames with public id instead.

**Comment by Cale111** [comment]
I made a fork with Steam Family Sharing support. [Cale111/PlayniteExtensions](https://github.com/Cale111/PlayniteExtensions)

I've noticed a bit of a problem though. It imports _everything,_ including programs like dedicated servers, and there's no option to disable those from being listed in the API. You can change `include_non_games` to false, but that also removes programs like Aseprite, which some people, including me, want to keep.

I also noticed it only reports games shared with you if the attribute `include_own` is _true._ Otherwise, it's only programs you own. I would've expected the opposite, but that's how it is.

You also need the Steam user's access token, rather than API key, which means another field in the add-on settings.

@EpiKnightz Maybe you were using the API wrong? It seems to work for me.

**Comment by EpiKnightz** [comment]
That great! When I tried it before it only returned non-private account game for me, maybe one of my steps was wrong. When I tried it just now it works fine.

I'm not well-verse with SteamAPI structure, but usually isn't Access Token only valid for a session though? Like it will be reset after a set of time or something, so you can't expect to input it once and be done with it.

**Comment by Cale111** [comment]
> That great! When I tried it before it only returned non-private account game for me, maybe one of my steps was wrong. When I tried it just now it works fine.
> 
> I'm not well-verse with SteamAPI structure, but usually isn't Access Token only valid for a session though? Like it will be reset after a set of time or something, so you can't expect to input it once and be done with it.
> 
> ![image](https://private-user-images.githubusercontent.com/1025684/351672791-032a6722-80aa-4e16-977c-70b5ba80b574.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4MTY2MTMsIm5iZiI6MTcyMTgxNjMxMywicGF0aCI6Ii8xMDI1Njg0LzM1MTY3Mjc5MS0wMzJhNjcyMi04MGFhLTRlMTYtOTc3Yy03MGI1YmE4MGI1NzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjRUMTAxODMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjhmZGY3ZjM2NzllYTIzNTU5MmY0OWEzOTYwZWYwNWY4YTAwNzNlNmU5YzNjM2Y1ZGJlMTE2YTBlYzdiNWU0OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.oDU6po7B_Z1e-CIp3ZkvQGx5QNI0KgZzNxQgEfoDDjI) Even the API site above show that it'll expire at some point.

Yes, it does expire pretty quickly, but it seems to be the only way to get the information. That's the only real problem now. It works well for one-time imports though!

The only way I can think of it auto-refreshing in Playnite is way beyond my abilities - perhaps it opens a web session to the Steam website in the background and copies the access token with some JS?

**Comment by GCTuba** [comment]
This would be a nice feature. The workaround I use is just installing games from my family's libraries that look interesting to me, updating the library in Playnite, then uninstalling the game. It will add the installed game to your library. A little annoying but it works.

**Comment by Bazzu85** [comment]
> This would be a nice feature. The workaround I use is just installing games from my family's libraries that look interesting to me, updating the library in Playnite, then uninstalling the game. It will add the installed game to your library. A little annoying but it works.

I tried too and it worked but the play time didn't sync with steam.
@EpiKnightz do you plan to update the plugin to support family games import?

thx for your work!

**Comment by EpiKnightz** [comment]
@JosefNemec to answer.
My solution is just a quick workaround, using additional account settings with API key and trying to get public library games if failed. It's not a proper fix so I probably won't make a pull request for it:
![image](https://github.com/user-attachments/assets/0e9007a5-194b-4ed5-8133-416f815d526c)

Though this fits my use case very well, only need to setup once and every time my family gets new games they will be added to my library. It's very convenient.

If Playnite can get access tokens from Steam and request family-shared games via API like @Cale111 implementation, that would be the best solution.

**Comment by cazzatedepa** [comment]
@Cale111 Which Playnite version should I use? I tried the latest but it's not finding "ProcessMonitor"

**Comment by Glitxhhh** [comment]
@Cale111 how do I add that fork to my playnite?

**Comment by Cale111** [comment]
> @Cale111 how do I add that fork to my playnite?

You have to download the forked version of the extension I linked, compile it in something like Visual Studio, then add the build directory that has the resulting .dll file to Playnite in the "For developers" menu.

Keep in mind this version is not very polished or anything. You have to replace the access token in the extension's settings almost every day or two. I just do it every few weeks when I decide to import, because games aren't added to my library often.

I use this URL to get my access token: https://store.steampowered.com/pointssummary/ajaxgetasyncconfig

It's the long string of text in the webapi_token key. Don't include the quotation marks.

**Comment by sena-nana** [comment]
I created a fork that uses the steam local cache appinfo to recognize all the games it can (including family library shares). I don't think it's a legitimate solution, so I won't be initiating PR, but enough for my own use. It's stable and doesn't require additional tokens (unless the format of appinfo is updated).
I'm a C# beginner (one day actually) so don't really have any code specification habits. Please forgive me if there are any problems with this.

**Comment by tamodolo** [comment]
> I created a fork that uses the steam local cache appinfo to recognize all the games it can (including family library shares). I don't think it's a legitimate solution, so I won't be initiating PR, but enough for my own use. It's stable and doesn't require additional tokens (unless the format of appinfo is updated). I'm a C# beginner (one day actually) so don't really have any code specification habits. Please forgive me if there are any problems with this.

This could be considered as another option to import steam games instead of login into steam at all. Games from ubisoft account are imported this way and don't need any login because of that. So I think this could be added to playnite as a way to import family shared games without the need of configuring another 4 or 5 acc and registering API keys.

So consider submiting it as I see this as a legitimate solution. Even if the PR isn't aproved imediately the chances are someone will improve and merge it. This is a shared pain in the arse of many of us. Seeing a solution to this will evoke fireworks.

**Comment by Kyrela** [comment]
> You'll have to add new stuff manually but it's a start anyway, with minimal manual labor

Thanks for this addition! It makes this process a lot easier.

**Comment by Antaneyes** [comment]
I created this addon and it apparently works perfectly to add the Steam Family games. I'm sharing it in case anyone is interested and it might be useful.
https://github.com/Antaneyes/SteamFamilyLibrary
